### ISSUE TITLE
Possible wake lock bug

### ISSUE THREAD

**Issue by jasongrishkoff** [Guichaguri/react-native-track-player#21]
I've seen the app crash a couple times when starting a song -- managed to grab a log. Looks like it has something to do wake lock?
```
06-11 15:47:28.972 30812 31653 E AndroidRuntime: FATAL EXCEPTION: mqt_native_modules
06-11 15:47:28.972 30812 31653 E AndroidRuntime: Process: com.decoderhq.indieshuffle, PID: 30812
06-11 15:47:28.972 30812 31653 E AndroidRuntime: java.lang.RuntimeException: WakeLock under-locked track-player-wake-lock
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at android.os.PowerManager$WakeLock.release(PowerManager.java:1300)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at android.os.PowerManager$WakeLock.release(PowerManager.java:1270)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at guichaguri.trackplayer.logic.MediaManager.onPlayerStop(MediaManager.java:315)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at guichaguri.trackplayer.logic.MediaManager.onStop(MediaManager.java:169)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at guichaguri.trackplayer.player.Player.updateState(Player.java:213)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at guichaguri.trackplayer.player.players.ExoPlayer.onPlayerStateChanged(ExoPlayer.java:199)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at com.google.android.exoplayer2.ExoPlayerImpl.handleEvent(ExoPlayerImpl.java:326)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at com.google.android.exoplayer2.ExoPlayerImpl$1.handleMessage(ExoPlayerImpl.java:93)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at android.os.Handler.dispatchMessage(Handler.java:102)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at android.os.Looper.loop(Looper.java:154)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at com.facebook.react.bridge.queue.MessageQueueThreadImpl$3.run(MessageQueueThreadImpl.java:208)
06-11 15:47:28.972 30812 31653 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:761)
06-11 15:47:28.977  4888  4906 W ActivityManager:   Force finishing activity com.decoderhq.indieshuffle/.MainActivity

```

**Comment by Guichaguri** [comment]
That seems to be related to #20. Same strange issue happening here

**Comment by jasongrishkoff** [comment]
It seems to be when I resume the app after some time away and start a new song.

**Comment by jasongrishkoff** [comment]
Not sure if this sheds some light on it? https://stackoverflow.com/questions/12140844/java-lang-runtimeexception-wakelock-under-locked-c2dm-lib

**Comment by Guichaguri** [comment]
I've looked that before, that's not the issue here.
For some reason, the service is being destroyed when starting a song, which is similar to #20.
I'll have to analyse my whole code trying to find what's wrong, and I'll probably get the Android music service examples to compare the code

**Comment by Guichaguri** [comment]
I've decided to change the whole code that links the module's functions with the service. Currently, I'm using a `Binder`, which should work according to the documentation, I don't know if that is what is causing the issue, but I'll be changing it to a `MediaController`

**Comment by jasongrishkoff** [comment]
Ah okay, let me know if I can help.  Hoping to take my app out of beta soon :)

**Comment by Guichaguri** [comment]
@jasongrishkoff There might be more API changes for this. The `MediaController` only allows one player to be "controlled". There are three things I can do:
* Remove support for multiple players, keeping only the main player
* Add a `createMainPlayer`, which would create/get the player that can play in background, and changing the behavior of `createPlayer`, which would create "simple" players that can only play while the app is in foreground (similar to all other react native modules)
* Keep the "multiple players" feature, using custom commands through the `MediaController`

Which one do you think would be the best option?

**Comment by jasongrishkoff** [comment]
1. So, the main reason to have multiple players is if someone is building a game or something and needs to fire off multiple sound files, right?  But this is "react native track player," and there's rarely a reason you'd want to play two songs at once (unless you're some mixing app).  So, the pros of keeping only the main player are that it's way simpler; the cons are that you can't play multiple sounds at once.  So I think the decision needs to be made as to what you want this library to be. Personally, I'd only ever need to play one sound at once because this is an audio streaming app I'm building, not a game.
2. Option 2 seems like a compromise that would allow main audio function, but with in-app sounds for interaction? Seems like it allows a broader subset to use the app, but adds complexity both from a code and usage standpoint.
3. Not entirely sure I understand the third one.

So yeah, in my head the first one sounds best.

**Comment by Guichaguri** [comment]
You're right, the only reason for multiple players is a mixing app (or a music app with song transitions, but this can be implemented without multiple players anyway) because games should use another type of player on Android (which is called `SoundPool`, made for short sounds).

It's definitely easier to maintain both the library and an app using it without having to hold the player ID.

The third option was "keeping everything the way they are, but just changing a few things internally", I'll think if there's really a good reason to go with option 3

**Comment by jasongrishkoff** [comment]
In the interim, maybe I'll wrap the wakeLock.release() like so to see if it helps:

```
if (wakeLock.isHeld()) {
    wakeLock.release()
}
```

**Comment by jasongrishkoff** [comment]
Closed with d7f1902f0d5a9c7f5f4e2c99779604868af686a1
